---
title: "per sample"
output: 
  html_document:
    css: sixicustom.css
    toc: true
    toc_float: 
      collapsed: true      
      smooth_scroll: true  
    number_sections: true
    self_contained: false
---

# Sample QC 

<details>

<summary>**The main tasks of sample-level QC include:**</summary>

-   Detecting technical artifacts that may caused by library construction, insufficient sequencing depth, amplification biases, and contaminations.
-   Identifying outlier samples with unexpected cell type compositions, et.al.
-   Recognizing incorrect sample labeling that might caused by errors in sample handling.

</details>

<details>

<summary>**To address these issues, the Sample QC module in SingleCellMQC includes three main components:**</summary>

-   [**Sample quality assessment -**](#md1-sec1) To check if there are any abnormalities in a variety of technical and biology metrics.
-   [**Outlier sample detection -**](#md1-sec2) To evaluate deviations in cell type composition.
-   [**Sample identity verification -**](#md1-sec3) To check whether the file labeling is consistent with analyzed data features.
</details>


------------------------------------------------------------------------

```{r Sample quality assessment setup, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if(length(data$Sample$Section1$Metrics )!=0){
  cat("
## Sample quality assessment {#md1-sec1}

**The main include: **

- **10X metrics:** Derived from the Cell Ranger output (if available).
- **Basic metrics:** Calculated by SingleCellMQC after preprocessing with tools such as Cell Ranger, dnbc4tools, and others.
- **V(D)J overview:** Detailed metrics and analysis of V(D)J data (if available).

------------------------------------------------------------------------
")
  cat('\n\n')
}


```



```{r cellranger warning, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if (!is.null(data$Sample$Section1$Metrics$Metrics_10X$base) && 
    length(data$Sample$Section1$Metrics$Metrics_10X$base) > 0){
  
  cat("
### 10X Metrics 

This section provides a summary of the 10X metrics alerts. The alerts are generated based on the [10X Genomics CellRanger ](https://github.com/10XGenomics/cellranger) requirements. The alerts are divided into two categories: errors and warnings. **Errors** indicate that the sample may have serious problems, while **warnings** indicate that the sample may have potential problems. The details below: 
      ")
  cat('\n\n')
}


if(length(data$Sample$Section1$Alert )!=0){
  if(length(data$Sample$Section1$Alert$table$Error)!=0){
    message <- paste0(paste0("- ", length(data$Sample$Section1$Alert$table$Error)," Error Samples: ", paste0(data$Sample$Section1$Alert$table$Error, collapse = ", "), " \n"),
    paste0( "- ", length(data$Sample$Section1$Alert$table$Warning)," Warning Samples: ", paste0(data$Sample$Section1$Alert$table$Warning, collapse = ", "), ""),
    " \n"
    )
    warning_alert_orange(message)
    
  }
  subchunkify(data$Sample$Section1$Alert$interactive_table,seed="cellranger_alerts")
  cat("\n\n")
}

# Check and output 10X metrics if data exists
if (!is.null(data$Sample$Section1$Metrics$Metrics_10X$base) && 
    length(data$Sample$Section1$Metrics$Metrics_10X$base) > 0) {
  subchunkify(data$Sample$Section1$Metrics$Metrics_10X$base, seed="Metrics_10X")
  cat('\n\n')
}


```

------------------------------------------------------------------------


```{r metrics introduction, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
if (length(data$Sample$Section1$Metrics )!=0) {
  cat("
### Basic Metrics { .tabset}

This section provides sample metrics calculated by SingleCellMQC after preprocessing with tools such as Cell Ranger, dnbc4tools, and others. The metrics include basic information and descriptive statistics for cell QC metrics. To identify potential warning samples, SingleCellMQC employs the `FindSampleMetricsWarning` function, which applies Median Absolute Deviation (MAD) to flag samples deviating from general distributions of key QC metrics. By default, the metrics analyzed include: `nCount_RNA`, `nFeature_RNA`, `nCount_ADT`, `nFeature_ADT`, `percent.mt`, and `percent.isotype`. Details are provided below:

")
  cat('\n\n')
}
```


```{r metrics outlier, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if(length(data$Sample$Section1$outlier_metrics)!=0){
  if(length(data$Sample$Section1$outlier_metrics$table)!=0){
    table_data <- data$Sample$Section1$outlier_metrics$table
    cat_word <- do.call(c, lapply(names(table_data), function(x){
      paste0( "- ", length(table_data[[x]])," ", x," outlier samples: ", paste0(table_data[[x]],  collapse = ", "), " \n")
    }))
    warning_alert_orange( paste0("Metrics outlier results (MAD Statistics): \n\n ",paste0(cat_word, collapse = ""),   " \n",
         "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"))
    
 }
  cat("\n\n")
  subchunkify(data$Sample$Section1$outlier_metrics$interactive_table,seed="interactive_table_metrics_outlier")
  cat("\n\n")
}

```


```{r metrics information, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if(length(data$Sample$Section1$Metrics )!=0){
  # cat( paste0(data$Sample$Section1$Metrics$note, collapse = "  \n") )
  note_grey(paste0(data$Sample$Section1$Metrics$note, collapse = "  \n"))
  cat('\n\n')
  

# Check and output Basic information if data exists
if (!is.null(data$Sample$Section1$Metrics$count$base) && 
    length(data$Sample$Section1$Metrics$count$base) > 0) {
  cat(paste0("#### Basic information", " {.tabset .unnumbered}\n"))
  subchunkify(data$Sample$Section1$Metrics$count$base, seed="count base information")
  cat('\n\n')
}

if (!is.null(data$Sample$Section1$Metrics$count$statistics) && 
    length(data$Sample$Section1$Metrics$count$statistics) > 0) {
  subchunkify(data$Sample$Section1$Metrics$count$statistics, seed="count base information statistics")
  cat('\n\n')
}

# Check and output Statistical summary if data exists
stat_data <- data$Sample$Section1$Metrics$summary
if (!is.null(stat_data) && length(stat_data) > 0) {
  cat(paste0("#### Descriptive statistics of cell QC metrics by sample", " {.tabset .unnumbered}\n"))
  i=1000
  for (g in names(stat_data)) {
    if (!is.null(stat_data[[g]]$base) && length(stat_data[[g]]$base) > 0) {
      cat('##### ', g, "{.unnumbered}",' \n')
      subchunkify(stat_data[[g]]$base, seed=paste0("summary_",g))
      cat("\n\n")
    }
  }
}

    
    cat("\n\n")
  }


```



```{r vdj information, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
if(length(data$Sample$Section1$VDJ )!=0){
  
  cat("
### V(D)J overview { .tabset}
")
  cat('\n\n')
  cat("This section provides a summary of the V(D)J information. The information includes V(D)J chain, V(D)J subtype, V(D)J gene usage, clonal quantity, and CDR3 length. The V(D)J gene usage, clonal quantity, and CDR3 length analysis based on [scRepertoire](https://github.com/ncborcherding/scRepertoire) package. The details below: \n\n")
  stat_data <- data$Sample$Section1$VDJ
  i=1000
  
  note_grey( paste0(data$Sample$Section1$VDJ$note, collapse = "  \n") ) 
  cat('\n\n')
  
  for (g in c("V(D)J chain", "V(D)J subtype") ) {
      cat('#### ', g, "{.unnumbered}",  ' \n')
      subchunkify(stat_data[[g]]$base,seed=paste0("VDJ_",g))
      cat("\n\n")
      subchunkify(stat_data[[g]]$statistics,seed=paste0("VDJ_statistics_",g))
      cat("\n\n")
  }
  cat("\n\n")

}


##geneUsage
if( length(data$Sample$Section1$VDJ$geneUsage )!=0 ){
  vdj_data <- data$Sample$Section1$VDJ$geneUsage
  name <- names(vdj_data)
  cat(paste0("#### geneUsage", " {.tabset .unnumbered}\n"))

  for (i in name) {
    cat(paste0("##### ", i, " {.tabset .unnumbered}\n"))
    subchunkify(vdj_data[[i]],seed= paste0(i, "_geneUsage") )
    cat('\n\n')
  }
  cat('\n\n')
}


##clonalQuant
if( length(data$Sample$Section1$VDJ$clonalQuant )!=0 ){
  vdj_data <- data$Sample$Section1$VDJ$clonalQuant
  name <- names(vdj_data)
  cat(paste0("#### clonalQuant", " {.tabset .unnumbered}\n"))

  for (i in name) {
    cat(paste0("##### ", i, " {.tabset .unnumbered}\n"))
    subchunkify(vdj_data[[i]],seed= paste0(i, "_clonalQuant") )
    cat('\n\n')
  }
  cat('\n\n')
}


##CDR3
if( length(data$Sample$Section1$VDJ$CDR3 )!=0 ){
  vdj_data <- data$Sample$Section1$VDJ$CDR3
  name <- names(vdj_data)
  cat(paste0("#### CDR3", " {.tabset .unnumbered}\n"))

  for (i in name) {
    cat(paste0("##### ", i, " {.tabset .unnumbered}\n"))
    cat("![](", paste0(cell_dir, "/Sample_QC/VDJ/CDR3/",  paste0(i,"_CDR3.png")), ")")
    cat("<p align='middle' style='font-size: 16px; color: black;'>**Figure: ** The distribution of CDR3 length. </p> \n\n")
        cat(paste0( "**IF you want to see interactive plotly object for details, you can see it by click the link ", "[plotly_link](", paste0( "./SingleCellMQC/plot/Sample_QC/VDJ/CDR3/",  paste0(i, "_CDR3.html")), ")", ".**\n\n")) 
    cat('\n\n')
  }

  cat('\n\n')
}




```


```{r cell type composition outlier, echo=FALSE, results='asis', warning=FALSE, message=FALSE}


if(length(data$Sample$Section2$outlier_comp)!=0){
  
  cat("
## PCT outlier sample detection {#md1-sec2 .tabset}

**Outlier sample detection in SingleCellMQC focuses on identifying deviations in cell type compositions. ** Deviations in cell type composition within samples from the same experimental group are often indicative of potential technical issues, such as prolonged processing time, variations in storage conditions, experimental processing errors, and batch effects, all of which can compromise the integrity of the data. To address this issue, SingleCellMQC incorporates cell type composition analysis as a additional outlier detection approach. This section include two main approaches: 1) Common range outliers; 2) Inter-sample outliers.The details below:

------------------------------------------------------------------------
")
  cat('\n\n')
  comp_data <- data$Sample$Section2$outlier_comp
  
  ###warining message
  
  
  if(length(comp_data$com_table)!=0){
      table_data <- comp_data$com_table
      cat_word <- do.call(c, lapply(names(table_data), function(x){
        paste0( "- ", length(table_data[[x]])," ", x," outlier samples: ", paste0(table_data[[x]],  collapse = ", "), " \n")
      }))
      warning_alert_orange( paste0("Common range outliers: \n\n ",paste0(cat_word, collapse = ""),   " \n",
           "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"))
  }
  
  
  if(length(comp_data$name_ellipse)!=0){
    ellipse_pa <- paste0( "Inter-sample PCT outliers (Ellipse): \n\n ",
    paste0("- ", "", length(comp_data$name_ellipse)," cell type composition outlier samples:  ", paste0(comp_data$name_ellipse, collapse = ", "), ""),
    " \n"
    )
  }else{
    ellipse_pa <- NULL
  }
  
  if(length(comp_data$name_dbscan)!=0){
    DBSCAN_pa <- paste0( "Inter-sample PCT outliers (DBSCAN): \n\n ",
    paste0("- ", "", length(comp_data$name_dbscan)," cell type composition outlier samples: ", paste0(comp_data$name_dbscan, collapse = ", "), ""),
    " \n"
    )
  }else{
    DBSCAN_pa <- NULL
  }
  
  if(!is.null(ellipse_pa) | !is.null(DBSCAN_pa) ){
    warning_alert_orange(paste0(ellipse_pa,  " \n", DBSCAN_pa,  " \n", 
                                "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"))
  }
  
    cat(paste0("### Common range outliers", " {.tabset .unnumbered}\n"))
    cat("This section provides a summary of the common range outliers based on cell type composition. Outlier samples can be identified by comparing their cell type composition with our established reference cell type composition ranges when performing the `FindCommonPCTOutlier` function. The reference cell type composition ranges are established based on the 2299 single cell samples from [DISCO](https://www.immunesinglecell.org/) database. \n")
  cat('\n\n')

    if(length(comp_data$com_table)!=0){
      subchunkify(comp_data$com_interactive_table,seed="com_interactive_table")
    }
    cat("\n\n")
    subchunkify(comp_data$CellTypePCT,seed="CellTypePCT")
        
    
    
    cat("\n\n")
    cat(paste0("### Inter-sample outliers", " {.tabset .unnumbered}\n"))
    
    cat("A PCA-based outlier detection method using confidence ellipse or Density-Based Spatial Clustering of Applications with Noise (DBSCAN) to identify samples with unusual cell type proportions within sample groups was also provided in this section. Additionally, contributions of each cell type composition to PCs were calculated, providing further insights into potential outliers and ensuring the reliability of the dataset for downstream analysis. \n")
    cat("\n\n")
 
  subchunkify(comp_data$interactive_table,seed="outlier_comp_interactive_table")
  cat("\n\n")

  cat(paste0("#### Ellipse", " {.tabset .unnumbered} \n\n"))
  subchunkify(comp_data$pca_ellipse, seed="outlier_comp_ellipse", fig_width=8)
  cat("\n\n")

  cat(paste0("#### DBSCAN", " {.tabset .unnumbered} \n\n"))
  subchunkify(comp_data$pca_dbscan, seed="outlier_comp_dbscan", fig_width=8)
  
}





```





```{r Sample label verification, echo=FALSE, results='asis', warning=FALSE}

if(!is.null(data$Sample$Section3)){
  
  cat('## Sample identity verification {#md1-sec3}

To avoid potential recording mistakes and data transmission errors, checking whether the file labeling is consistent with analyzed data features is necessary at the beginning of sample QC. In this section, you can check if your sample gender information is incorrectly labeled. In addition, if your sample does not have gender information, you can get it according to the table below.')
  cat('\n\n')

  
identity <- data$Sample$Section3$identity
subchunkify(reactablefmtr::add_source(identity, "Table: The predicted sex information"), seed="identity")


 note_grey(paste0("> **NOTE:**  \n>  \n> - **XIST** is female-expressed from its function in X-chromosome inactivation. **DDX3Y**, **UTY** and **RPS4Y1** are Y-chromosome-specific. **`Predict`** column represents the analyzed sex.  \n", '> -  The analyzed sex results: **Female: ', sum(data$Sample$Section3$num %in% "Female") , " ;  Male: ", sum(data$Sample$Section3$num %in% "Male"), " ;**"))
cat("\n\n")
}


```

<hr>



